pkg/kubectl/cmd/util/factory_object_mapping.go

```diff
@@ -489,3 +490,95 @@ func (f *ring1Factory) OpenAPISchema(cacheDir string) (openapi.Resources, error)
 	// Delegate to the OpenAPIGetter
 	return f.openAPIGetter.getter.Get()
 }
+
+func (f *ring1Factory) ValidResourcesFromDiscoveryClient() string {
+	resourceMap := make(map[string]metav1.APIResource)
+	var keys []string
+	discoveryClient, err := f.clientAccessFactory.DiscoveryClient()
+	if err == nil {
+		apiResList, err := discoveryClient.ServerResources()
+		if err == nil {
+			for _, apiResources := range apiResList {
+				for _, apiRes := range apiResources.APIResources {
+					if !strings.Contains(apiRes.Name, "/") && apiRes.Name != "apiservices" {
+						if _, ok := resourceMap[apiRes.Name]; !ok {
+							resourceMap[apiRes.Name] = apiRes
+							keys = append(keys, apiRes.Name)
+						}
+					}
+				}
+			}
+			keys = append(keys, "clusters")
+			sort.Strings(keys)
+			row := "Valid resource types include:\n\n  * all\n"
+			for _, k := range keys {
+				if k == "clusters" {
+					row = row + "  * clusters (valid only for federation apiservers)\n"
```

Comment:
I'd do this purely on discovery.  If they're in the discovery list, they don't need qualification